lib/error.g

```diff
@@ -189,21 +189,50 @@ BIND_GLOBAL("ErrorInner",
     ERROR_COUNT := ERROR_COUNT+1;
     errorLVars := ErrorLVars;
     ErrorLVars := context;
-    # BreakOnError is defined by the `-T` command line flag in init.g
-    if QUITTING or not BreakOnError then
-        if not SilentErrors then
-            PrintTo("*errout*", "Error, ");
-            for x in earlyMessage do
-                PrintTo("*errout*", x);
-            od;
-            PrintTo("*errout*", "\n");
+    # The variables BreakOnError and QUITTING determine whether we
+    # start an interactive break loop or whether we use JUMP_TO_CATCH
+    # to exit this function.
+    # BreakOnError is defined by the `-T` command line flag in init.g.
+    doBreakLoop := BreakOnError and not QUITTING;
+    # If we do not start a break loop, the variables SilentErrors and
+    # AlwaysPrintTracebackOnError determine whether any messages or
+    # respectively tracebacks are printed.
+    # SilentErrors is used by HPC-GAP to e.g. suppress error messages in
+    # worker threads.
+    if doBreakLoop or not SilentErrors then
+        PrintTo("*errout*", "Error, ");
+        for x in earlyMessage do
+            PrintTo("*errout*", x);
+        od;
+        if doBreakLoop or AlwaysPrintTracebackOnError then
+            if printThisStatement then
+                if context <> GetBottomLVars() then
+                    PrintTo("*errout*", " in\n  ");
+                    PRINT_CURRENT_STATEMENT("*errout*", context);
+                fi;
+            else
+                location := CURRENT_STATEMENT_LOCATION(context);
+                if location <> fail then
+                  PrintTo("*errout*", " at ", location[1], ":", location[2]);
+                fi;
+            fi;
+            PrintTo("*errout*", " called from");
         fi;
+        PrintTo("*errout*", "\n");
+    fi;
+    if IsBound(OnBreak) and IsFunction(OnBreak)
+        and (doBreakLoop or (not SilentErrors and AlwaysPrintTracebackOnError)) then
+        OnBreak();
+    fi;
+    # Do we want to skip the break loop?
+    if not doBreakLoop then
         if IsHPCGAP then
             LastErrorMessage := "";
             lastErrorStream := OutputTextString(LastErrorMessage, true);
             for x in earlyMessage do
                 PrintTo(lastErrorStream, x);
             od;
+            # FIXME: Also make HPCGAP work with AlwaysPrintTracebackOnError
```

Comment:
What needs to be done to achieve this? I see some code added for HPC-GAP, is this comment obsolete?

Reply:
This isn't obsolete. IMO if `AlwaysPrintTracebackOnError` is true, the traceback should also be put into `LastErrorMessage`. To do this there needs to be a way to put the output of `Where()` into `lastErrorStream`.

I'll add a comment to the source.